mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala

```diff
@@ -0,0 +1,201 @@
+   * @group param
+   */
+  val epsilon = new Param[Double](this, "epsilon", "distance threshold")
+
+  /** @group getParam */
+  def getEpsilon: Double = $(epsilon)
+
+  /**
+   * Param for the initialization algorithm. This can be either "random" to choose random points as
+   * initial cluster centers, or "k-means||" to use a parallel variant of k-means++
+   * (Bahmani et al., Scalable K-Means++, VLDB 2012). Default: k-means||.
+   * @group param
+   */
+  val initMode = new Param[String](this, "initMode", "initialization algorithm",
+    (value: String) => MLlibKMeans.validateInitializationMode(value))
+
+  /** @group getParam */
+  def getInitializationMode: String = $(initMode)
+
+  /**
+   * Param for the number of steps for the k-means|| initialization mode. This is an advanced
+   * setting -- the default of 5 is almost always enough. Default: 5.
+   * @group param
+   */
+  val initSteps = new Param[Int](this, "initSteps", "number of steps for k-means||",
+    (value: Int) => value > 0)
+
+  /** @group getParam */
+  def getInitializationSteps: Int = $(initSteps)
+
+  /**
+   * Validates and transforms the input schema.
+   * @param schema input schema
+   * @return output schema
+   */
+  protected def validateAndTransformSchema(schema: StructType): StructType = {
+    SchemaUtils.checkColumnType(schema, $(featuresCol), new VectorUDT)
+    SchemaUtils.appendColumn(schema, $(predictionCol), IntegerType)
+  }
+}
+
+/**
+ * :: Experimental ::
+ * Model fitted by KMeans.
+ *
+ * @param parentModel a model trained by spark.mllib.clustering.KMeans.
+ */
+@Experimental
+class KMeansModel private[ml] (
+    override val uid: String,
+    private val parentModel: MLlibKMeansModel) extends Model[KMeansModel] with KMeansParams {
+
+  override def copy(extra: ParamMap): KMeansModel = {
+    val copied = new KMeansModel(uid, parentModel)
+    copyValues(copied, extra)
+  }
+
+  override def transform(dataset: DataFrame): DataFrame = {
+    val predictUDF = udf((vector: Vector) => predict(vector))
+    dataset.withColumn($(predictionCol), predictUDF(col($(featuresCol))))
+  }
+
+  override def transformSchema(schema: StructType): StructType = {
+    validateAndTransformSchema(schema)
+  }
+
+  private[clustering]
+  def predict(features: Vector): Int = parentModel.predict(features)
+
+  def clusterCenters: Array[Vector] = parentModel.clusterCenters
+}
+
+/**
+ * :: Experimental ::
+ * KMeans API for spark.ml Pipeline.
+ */
+@Experimental
+class KMeans(override val uid: String) extends Estimator[KMeansModel] with KMeansParams {
+
+  setDefault(k, 2)
+  setDefault(maxIter, 20)
+  setDefault(runs, 1)
+  setDefault(initMode, MLlibKMeans.K_MEANS_PARALLEL)
+  setDefault(initSteps, 5)
+  setDefault(epsilon, 1e-4)
+  setDefault(seed, Utils.random.nextLong())
+
+  override def copy(extra: ParamMap): Estimator[KMeansModel] = defaultCopy(extra)
+
+  def this() = this(Identifiable.randomUID("kmeans"))
+
+  /** @group setParam */
```

Comment:
Those setters in GBTParams work with Java because the subclasses override them.  I probably should not have done that; it seems safer not to have setters in the trait since that risks developers carelessly not adding overridden setters to the concrete classes.

At any rate, the setters for this should be in the concrete classes.  I think it rarely makes sense to move them elsewhere.  We probably should have a setter in both the Estimator and Model classes since users (or meta-algorithms) might want to modify the featuresCol of the Model.